ports/stm32/boards/NUCLEO_WB55/rfcore_makefirmware.py

```diff
@@ -41,27 +44,149 @@
+    addr_640k = int(addr_640k, 0)
+    addr_512k = int(addr_512k, 0)
+    addr_256k = int(addr_256k, 0)
+    version = [int(v) for v in version.lower().lstrip(b"v").split(b".")]
+    return addr_1m, addr_640k, addr_512k, addr_256k, version
 
 
 def main(src_path, dest_path):
-    for src_file, dest_file in _FIRMWARE_FILES.items():
-        src_file = os.path.join(src_path, src_file)
+
+    # Load the release note to parse for important details
+    with open(os.path.join(src_path, _RELEASE_NOTES), "rb") as f:
+        release_notes = f.read()
+    # remove some formatting
+    release_notes = re.sub(rb"</?strong>", b"", release_notes)
+    # flatten tables
+    release_notes = re.sub(
+        rb"</t[dh]>\W*\n*\W*",
+        b",",
+        release_notes.replace(b"<td>", b"").replace(b"<th>", b""),
+    )
+    if (
+        b"Wireless Coprocessor Binary,STM32WB5xxG(1M),STM32WB5xxY(640k),STM32WB5xxE(512K),STM32WB5xxC(256K),Version,"
+        not in release_notes
+    ):
+        print(
+            "Cannot determine binary load address, please confirm manually in Release Notes."
+        )
+        release_notes = None
+
+    for src_filename, dest_file in _FIRMWARE_FILES.items():
+        src_file = os.path.join(src_path, src_filename)
         dest_file = os.path.join(dest_path, dest_file)
         if not os.path.exists(src_file):
             print("Unable to find: {}".format(src_file))
             continue
         sz = 0
         with open(src_file, "rb") as src:
+            crc = 0
             with open(dest_file, "wb") as dest:
                 while True:
                     b = src.read(4)
                     if not b:
                         break
                     (v,) = struct.unpack("<I", b)
                     v ^= _OBFUSCATION_KEY
-                    dest.write(struct.pack("<I", v))
+                    vs = struct.pack("<I", v)
+                    dest.write(vs)
+                    crc = crc32(vs, crc)
                     sz += 4
-        print("Written {} ({} bytes)".format(dest_file, sz))
+
+                addr_1m, addr_640k, addr_512k, addr_256k, version = get_details(
+                    release_notes, src_filename
+                )
+
+                footer = struct.pack(
+                    "<37sIIIIbbbI",
+                    src_filename.encode(),
+                    addr_1m,
+                    addr_640k,
+                    addr_512k,
+                    addr_256k,
+                    *version,
+                    _OBFUSCATION_KEY,
+                )
+                assert len(footer) == 60
+                dest.write(footer)
+                crc = crc32(footer, crc)
+                crc = 0xFFFFFFFF & -crc - 1
+                dest.write(struct.pack("<I", crc))
+                sz += 64
+
+        print(
+            f"Written {src_filename} v{version[0]}.{version[1]}.{version[2]} to {dest_file} ({sz} bytes)"
+        )
+        check_file_details(dest_file)
+
+
+def validate_crc(f):
```

Comment:
I'm not sure it's that important but would be good to avoid duplication... We could put these in `rfcore_firmware_utils.py` and make sure both get frozen?

Reply:
rfcore_firmware.py could be refactored slightly to allow it to be imported into rfcore_makefirmware, then it could be single source of truth for a few things (this function, the KEY, etc)